File: PokeIn/PokeIn_Free_v2.032/PokeIn_Free_v2.032/VB.NET ChatSample/ChatSampleVB/ChatApp.vb
Imports PokeIn
Imports PokeIn.Comet

Public Class ChatApp
	Implements IDisposable
	
    Private ClientID As String
    Public Shared Names As Dictionary(Of String, String)
    Private Username As String
    Public Shared Users As Dictionary(Of String, String)

    Public Overloads Sub Dispose() Implements IDisposable.Dispose
       'Do whatevet you want here, PokeIn will call this method when a user is disconnected
	   Try
            SyncLock ChatApp.Users
                ChatApp.Users.Remove(Me.ClientID)
            End SyncLock
            SyncLock ChatApp.Names
                ChatApp.Names.Remove(Me.Username)
            End SyncLock
        Finally
            MyBase.Finalize()
        End Try
    End Sub

    Public Sub New(ByVal clientId As String)
        Me.ClientID = clientId
        Me.Username = ""
    End Sub

    Shared Sub New()
        ChatApp.Users = New Dictionary(Of String, String)
        ChatApp.Names = New Dictionary(Of String, String)
    End Sub

    Protected Overrides Sub Finalize()

    End Sub

    Public Sub Send(ByVal message As String)
        Dim json As String = PokeIn.JSON.Method("ChatMessageFrom", Me.Username, message) ' ChatMessageFrom( "user", "message" );

        CometWorker.SendToAll(json)
    End Sub

    Public Sub SetName(ByVal user_name As String)
        If (Me.Username <> "") Then
            CometWorker.SendToClient(Me.ClientID, "alert('You already have an username!');")
        Else
            Dim duplicate As Boolean = False
            SyncLock ChatApp.Names
                duplicate = ChatApp.Names.ContainsKey(user_name)
            End SyncLock
            If duplicate Then
                CometWorker.SendToClient(Me.ClientID, "alert('Another user is using the name you choose!\nPlease try another one.');")
            Else
                SyncLock ChatApp.Names
                    ChatApp.Names.Add(user_name, Me.ClientID)
                End SyncLock
                SyncLock ChatApp.Users
                    ChatApp.Users.Add(Me.ClientID, user_name)
                End SyncLock
                Me.Username = user_name
                Dim js As String = JSON.Method("UsernameSet", user_name)

                CometWorker.SendToClient(Me.ClientID, js)
            End If
        End If
    End Sub

End Class

File: PokeIn/PokeIn_Free_v2.032/PokeIn_Free_v2.032/Chat/ChatApp.cs
using System;
using System.Collections.Generic;
using PokeIn;
using PokeIn.Comet;

namespace ChatSample
{
    //Our custom parameter class
    [Serializable]
    public class ChatMessage
    {
        public string Username, Message;

        //You should define an Empty Consturctor to allow PokeIn to de-serialize
        public ChatMessage()
        {
            Username = "";
            Message="";
        }

        public ChatMessage(string userName, string message)
        {
            Username = userName;
            Message = message;
        }
    }

    //Chat Class
    public class ChatApp:IDisposable
    {
        public static Dictionary<string, string> Users = new Dictionary<string,string>();
        public static Dictionary<string, string> Names = new Dictionary<string,string>();
        string _clientId;
        string _username;
        public ChatApp(string clientId)
        {
            _clientId = clientId;
            _username = "";
        }
        public void Dispose() 
        {
            Send(new ChatMessage(_username, "Disconnected"));
            lock (Users)
            {
                Users.Remove(_clientId);
            }
            lock (Names)
            {
                Names.Remove(_username);
            } 
        } 
        public void SetName(string userName)
        {
            if (_username != "")
            {
                CometWorker.SendToClient(_clientId, "alert('You already have an username!');btnChat.disabled = '';");
                return;
            }
            bool duplicate;
            lock (Names)
            {
                duplicate = Names.ContainsKey(userName);
            }
            if (duplicate)
                CometWorker.SendToClient(_clientId, "alert('Another user is using the name you choose!\\nPlease try another one.');btnChat.disabled = '';");
            else
            {
                lock (Names)
                {
                    Names.Add(userName, _clientId);
                }

                lock (Users)
                {
                    Users.Add(_clientId, userName);
                }
                _username = userName;

                //Create JSON Method
                string json = JSON.Method("UsernameSet", userName);//UsernameSet('UserName');
                CometWorker.SendToClient(_clientId, json); 
            }
        }

        //When PokeIn sees ChatMessage custom class as a parameter, it automaticly defines ChatMessage JS class on client side.
        public void Send(ChatMessage message) 
        {
            //Create JSON method from custom class
            string json = JSON.Method("ChatMessageFrom", message); //ChatMessageFrom( {Username:'username', Message:'message' } );
            CometWorker.SendToAll(json);
        }
    }
}
